### DISCUSSION TITLE
Arctic magnetometer problems

### DISCUSSION THREAD

**Discussion by harrylerman** [dannyjacobs/ECHO#54] [General]
So at MARS, the magnetometer seemed to be unhappy. This was evident in two ways, first with the drone just sitting on the ground and not moving, its heading was drifting by 40 degrees (I saw it going from 170 to 210 erratically). The second is a mavlink message parameter called the "magnetometer innovation test ratio." I spent a while looking into what this parameter means, like most drone stuff there's not a lot published but I think I understand it.

People smart in the world of signal processing might already know this, but in signal processing and particularly in the land of Kalman filters (https://en.wikipedia.org/wiki/Kalman_filter) which the drone uses for its state estimation (https://docs.px4.io/v1.12/en/advanced_config/tuning_the_ecl_ekf.html), "innovation" has a particular meaning. Here's a wiki link about it: https://en.wikipedia.org/wiki/Innovation_(signal_processing). It's similar to a residual, it is the difference between the predicted state at a time t and the actual measured state at time t.

> the innovation is the difference between the observed value of a variable at time t and the optimal forecast of that value based on information available prior to time t.
> In contrast, the residual is the difference between the observed value of a variable at time t and the optimal updated state of that value based on information available till (including) time t.

The drone calculates its magnetometer innovation and then computes a ratio of the innovation to an "acceptable" level. These quantities are in expressed in terms of standard deviations, and the acceptable level which is called the "mag_gate" parameter (ctrl-f it here: https://docs.px4.io/main/en/advanced_config/parameter_reference.html) (which is configurable) is 3 standard deviations by default. The mag ratio should be 0.5 or below at all times (https://mavlink.io/en/messages/common.html#ESTIMATOR_STATUS):

> Under normal operation the innovation test ratios should be below 0.5 with occasional values up to 1.0. Values greater than 1.0 should be rare under normal operation and indicate that a measurement has been rejected by the filter. The user should be notified if an innovation test ratio greater than 1.0 is recorded. Notifications for values in the range between 0.5 and 1.0 should be optional and controllable by the user.

Up at MARS the ratio was far above 0.5, it was up to about 4. Here is a plot of the mag_ratio parameter from the long unloaded hover test in May:
![may11mag](https://user-images.githubusercontent.com/66405902/183500178-fe158676-c5d3-439d-82ba-50020d1ea50b.png)

and then a plot of that parameter at MARS:
![marsmag](https://user-images.githubusercontent.com/66405902/183500200-4f76501d-83bf-45e1-8bd2-3c960fbf2d22.png)

(y-axis is the mag_ratio and x-axis is Unix time so effectively time in seconds. I realize better axis labels could be used, these are just quick plots for now). Seems apparent that the magnetometer at MARS was out to lunch!

**Comment by dannyjacobs** [comment]
Interesting. Presumably this is happening because the XY component of the field is much lower, because the field vector is nearly vertical to the ground. Are there any calibrated field strength measurements from MARS or around there?  I guess we could always consult a magnetic model.
The trace from the MARS test is not completely bad. There seems to be a good stretch where things are  <0.5. Though not much. Did you do anything during that trace? Like maybe spool up the motors?

**Reply by harrylerman** [reply]
I think it's definitely the magnetic field being practically vertical up at MARS. Some plots of the Earth's magnetic inclination are available here: https://www.ngdc.noaa.gov/geomag/WMM/image.shtml and where we were it is about 88 degrees, vs about 71 here in Montreal. So just 2 deg off of vertical up there. Probably not much signal for the magnetometer. Pixhawk does mention in its EKF tuning section that using a dual GPS setup for heading is probably necessary in locations with a high magnetic inclination. 

Here's data from the logs showing that mag ratio along with throttle - you can see the three (short) flights I did before the crash. Sort of hard to say if anything really correlates.  
![mars2](https://user-images.githubusercontent.com/66405902/184273121-0b50ac35-e4af-4967-bcc7-610ce9c42903.png)

Just to show it, here's the heading with that mag ratio. You can see a long stretch in the middle where the drone was just sitting stationary on the ground from about 7700s - 8600s, and the heading is bouncing between about 170 and 210. 
![MARS3](https://user-images.githubusercontent.com/66405902/184273210-4a147af7-c5b6-4bfb-b650-28327584a5b0.png)
